src/main/kotlin/am/ik/yavi/builder/ValidatorBuilderExtensions.kt

```diff
@@ -177,4 +177,34 @@ fun <T> ValidatorBuilder<T>.konstraintOnCondition(condition: ConstraintCondition
         this.constraintOnCondition(condition, ValidatorBuilder.ValidatorBuilderConverter { it.apply(block) })
 
 fun <T> ValidatorBuilder<T>.konstraintOnGroup(group: ConstraintGroup, block: ValidatorBuilder<T>.() -> Unit): ValidatorBuilder<T> =
-        this.constraintOnCondition(group.toCondition(), ValidatorBuilder.ValidatorBuilderConverter { it.apply(block) })
\ No newline at end of file
+        this.constraintOnCondition(group.toCondition(), ValidatorBuilder.ValidatorBuilderConverter { it.apply(block) })
+
+inline fun <T> validator(init: ValidatorBuilderKt<T>.() -> Unit): Validator<T> {
+        val builder = ValidatorBuilderKt<T>()
+        return builder.apply(init).build()
+}
+
+class ValidatorBuilderKt<T> : ValidatorBuilder<T>() {
+
+        /** Q: this will be required for each type, right? */
+        operator fun KProperty1<T, String?>.invoke(block: CharSequenceConstraint<T, String?>.() -> Unit) =
+                constraint(this, this.name) { it.apply(block) }
+
+        /** Q: this will be required for each type, right? */
+        infix fun KProperty1<T, String?>.required(block: CharSequenceConstraint<T, String?>.() -> Unit) =
+                constraint(this, this.name) { it.notNull().apply(block) }
+
+        infix fun <L: Collection<E>?, E> KProperty1<T, L?>.forEach(block: ValidatorBuilderKt<E>.() -> Unit) {
+                TODO()
+                /** can't get following to work due to cumbersome inheritance setup of ValidatorBuilderKt :-( */
+                // forEach(this, this.name) { it.apply(block) }
+        }
+
+        infix fun <N> KProperty1<T, N?>.nest(block: ValidatorBuilderKt<N>.() -> Unit) {
+                // TODO()
+                /** also not possible to cast... :-(( */
+                // nest(this, this.name) { it.apply(block as ValidatorBuilder<N?>.() -> Unit) }
+        }
```

Comment:
Would anyone be able to help unblock / suggest a way forward?
Are there any options without wider refactoring of Yavi itself? 
(and without moving/rewriting Yavi in Kotlin 😅)

Reply:
```kotlin
inline fun <T> validator(init: ValidatorBuilderKt<T>.() -> Unit): Validator<T> {
        val builder = ValidatorBuilderKt<T>(ValidatorBuilder())
        return builder.apply(init).build()
}

class ValidatorBuilderKt<T>(private val validatorBuilder: ValidatorBuilder<T>) {

        fun build(): Validator<T> {
            return validatorBuilder.build()
        }

        /** Q: this will be required for each type, right? */
        operator fun KProperty1<T, String?>.invoke(block: CharSequenceConstraint<T, String?>.() -> Unit) =
                validatorBuilder.constraint(this, this.name) { it.apply(block) }

        /** Q: this will be required for each type, right? */
        infix fun KProperty1<T, String?>.required(block: CharSequenceConstraint<T, String?>.() -> Unit) =
                validatorBuilder.constraint(this, this.name) { it.notNull().apply(block) }

        infix fun <L: Collection<E>?, E> KProperty1<T, L?>.forEach(block: ValidatorBuilderKt<E>.() -> Unit) {
                validatorBuilder.forEach(this, this.name) { ValidatorBuilderKt(it).apply(block).validatorBuilder }
        }

        infix fun <N> KProperty1<T, N>.nest(block: ValidatorBuilderKt<N>.() -> Unit) {
               validatorBuilder.nest(this, this.name) { ValidatorBuilderKt(it).apply(block).validatorBuilder }
        }

        infix fun <N> KProperty1<T, N?>.nest(validator: Validator<N>): ValidatorBuilder<T> = TODO()
}
```

Reply:
brilliant, thank you! ✨ 
looks like a solid pattern to proceed with...